cuqipy_fenics/pde.py

```diff
@@ -11,60 +11,133 @@
 ]
 
 class FEniCSPDE(PDE,ABC):
-    def __init__(self, PDE_form, mesh, solution_function_space, parameter_function_space, dirichlet_bc,observation_operator=None):
```

Comment:
Could you add a doc string here for the new parameters? Perhaps it will also explain what you mean by the parameter property below. In particular why you can set it. Similar for the forward solution, not sure what this means and why its reset when setting parameter?

I am having a bit of a hard time following the logic in the code for the subclassing classes also. Might help with more doc + comments in code.

Reply:
Thank you @nabriis for your very helpful review and edits! I added documentation for the classes and update the variables names J & J2 adding some comments about them. I also added a reference for adjoint based derivatives.

I appreciate it if you can take a look and let me know what do you think. 

I think using auto-styling is a great idea! and is probably overdue :). btw, I wanted to mention that in some of the style changes you added as suggestion, the indentation was off, it could be something not right with the GitHub suggested code feature?